Add memory system call replacements.

davidchisnall opened the pull request:
These system calls are used for heap setup and so we can't yet remove
the musl wrappers, but this provides the kernel support that we'll need
when we do.

davidchisnall committed:
Add memory system call replacements.

These system calls are used for heap setup and so we can't yet remove
the musl wrappers, but this provides the kernel support that we'll need
when we do.

The mmap override now needs to call the function that implements the
read system call directly, rather than going via the system call layer.
While modifying that code, also fix a bug where we were modifying the
offset of the file descriptor during mmap.

After this change, all of the functions, including the `enclave_*` ones,
use the syscall convention for reporting errors: a positive result
indicates success, a negative result indicates failure, with the error
code negated.

This change also updates the callers of these functions to correctly
check the results.  One case, in lkl/setup.c, was checking the return
value of `lkl_sys_mmap` against `MAP_FIXED`, so would only detect errors
that returned `-EPERM` and not any other kind of failure.  This
particular place happened sufficiently early that failure probably did
not happen.

prp commented on src/enclave/enclave_mem.c:
```diff
@@ -390,7 +270,7 @@ void* enclave_mmap(
         {
             // Set requested page permission
-            mprotect(ret, length, prot);
+			sgxlkl_host_syscall_mprotect(&mprotect_ret, ret, length, prot);
```
Is there some weird indentation here?

davidchisnall replied:
I forgot to run clang-format before I committed this.  Fixed now.

prp commented on src/include/enclave/enclave_mem.h:
```diff
@@ -29,13 +29,15 @@ void* enclave_mremap(
     int mremap_fixed);
 
-int enclave_mmap_flags_supported(int flags, int fd);
+extern int mmap_files; // Allow MAP_PRIVATE or MAP_SHARED?
```
Why not define `mmap_files` here?

davidchisnall replied:
And make it common linkage?  I try to avoid common linkage whenever possible because it causes all manner of subtle linker behaviours.

prp reviewed (approved):
LGTM

SeanTAllen commented on src/enclave/enclave_mem.c:
```diff
@@ -98,98 +89,6 @@ static size_t addr_to_index(void* addr)
-        if (mem > 0)
-        {
-            // Read file into memory
-            lseek(fd, offset, SEEK_SET);
```
Did you intend to remove this lseek when this function was moved and partially rewritten?

davidchisnall replied:
Yes.  lseek changes the offset in the fd, which is something mmap should not do.

SeanTAllen replied:
David replied elsewhere that it was intentional

SeanTAllen commented:
I'm working on this PR now.

Commit:
Fix incorrect flag in mmap fallback logic

Commit:
Remove incorrect munmap03 patch

There is a bug in munmap. It should return -1 on an error, not 0.
The patch was attempting to work around the bug rather than doing
what the test should do, fail and report the bug.

Commit:
Remove bad munmap03 LTP test patch

The patch is wrong. It hides a bug in munmap. On failure, munmap is
supposed to return -1. The patch as noted in its comment allows for
0 aka success to be returned and assumes because the errno was set
correctly, that everything is correct.

It isn't. The return value has to be -1 when failure occurs. Anyone
using this munmap interface would get surprising results and the test
patch hides that. munmap is currently broken.

Commit:
Fix failing munmap03 testFix failing munmap03 test

Commit:
Return long from memory syscall functions

If you don't, for reasons I haven't looked into yet, the errno translation
of taking something like -22 and changing to -1 return value with an errno
of 22 doesn't work.

Commit:
Bump to latest sgx-lkl-musl

Commit:
Didn't mean to commit that

SeanTAllen commented:
This is currently blocked waiting for #577 to land. In the meantime, I've done some related work in #655.

Once #655 and #577 are merged to oe_port, I'm going to pick up the work on this again and remove the syscall overrides for mmap functions from sgx-lkl-musl and into system call replacements like this PR has.

Some of the changes needed to make this change possible are part of #655.

SeanTAllen commented:
Closed in favor of #710. I'm leaving the branch around until #710 is merged.
